File: 백준/Gold/2206. 벽 부수고 이동하기/벽 부수고 이동하기.js
class Queue {
  constructor() {
    this.queue = [];
    this.head = 0;
    this.tail = 0;
  }

  push(item) {
    this.queue[this.tail++] = item;
  }

  pop() {
    return this.queue[this.head++];
  }

  isEmpty() {
    return this.head === this.tail;
  }
}

const input = require("fs")
  .readFileSync(0)
  .toString()
  .trim()
  .split("\n")
  .map((v) => v.trim());

const [N, M] = input[0].split(" ").map(Number);
const map = input.slice(1).map((v) => v.split("").map(Number));
const visited = Array.from({ length: N }, () =>
  Array.from({ length: M }, () => Array(2).fill(false))
);

const q = new Queue();
q.push([0, 0, 1, 0]); // x, y, dist, broken
visited[0][0][0] = true;

let answer = -1;

const dx = [-1, 0, 1, 0];
const dy = [0, -1, 0, 1];

while (!q.isEmpty()) {
  const [x, y, dist, broken] = q.pop();

  if (y === N - 1 && x === M - 1) {
    answer = dist;
    break;
  }

  for (let i = 0; i < 4; i++) {
    const nx = x + dx[i];
    const ny = y + dy[i];

    if (nx < 0 || nx >= M || ny < 0 || ny >= N) continue;

    if (map[ny][nx] === 1 && broken === 0) {
      if (!visited[ny][nx][1]) {
        visited[ny][nx][1] = true;
        q.push([nx, ny, dist + 1, 1]);
      }
    }

    if (map[ny][nx] === 0) {
      if (!visited[ny][nx][broken]) {
        visited[ny][nx][broken] = true;
        q.push([nx, ny, dist + 1, broken]);
      }
    }
  }
}

console.log(answer);

File: 백준/Gold/2206. 벽 부수고 이동하기/벽 부수고 이동하기.py
import sys
from collections import deque

input_data = sys.stdin.read().split()

[N, M] = [int(input_data[0]), int(input_data[1])]
board = []

idx = 2
for _ in range(N):
    board.append(list(map(int, list(input_data[idx]))))
    idx += 1

# visited[r][c][0]: 벽 안 부숨, [1]: 벽 부숨
visited = [[[False] * 2 for _ in range(M)] for _ in range(N)]
q = deque([(0, 0, 1, 0)]) # r, c, dist, broken
visited[0][0][0] = True

dx = [0, 0, -1, 1]
dy = [-1, 1, 0, 0]
answer = -1

while q:
    y, x, dist, broken = q.popleft()

    if y == N - 1 and x == M - 1:
        answer = dist
        break

    for i in range(4):
        ny, nx = y + dy[i], x + dx[i]

        if 0 <= ny < N and 0 <= nx < M:
            # 벽을 만났을 때
            if board[ny][nx] == 1 and broken == 0:
                if not visited[ny][nx][1]:
                    visited[ny][nx][1] = True
                    q.append((ny, nx, dist + 1, 1))
            
            # 빈 공간일 때
            elif board[ny][nx] == 0:
                if not visited[ny][nx][broken]:
                    visited[ny][nx][broken] = True
                    q.append((ny, nx, dist + 1, broken))

print(answer)